### ISSUE TITLE
point/pixel-wise normalization seems not supported by current DALI

### ISSUE THREAD

**Issue by JanuszL** [NVIDIA/DALI#43]
Point/pixel-wise normalization seems not supported by current DALI. DALI’s normalization operators receive channel-wise means and stds (in other words, three means and three stds), but does not receive point-wise means and stds.

**Comment by JanuszL** [comment]
Internal tracker DALI-148

**Comment by paulbisso** [comment]
It would also be useful to rescale pixels to a [0,1] range. Pytorch's pretrained models, for instance, presume floating-point pixel values between [0,1]. I've gotten around this by multiplying the 'mean' and 'std' inputs by 255 manually, but it would make for a more smooth and less error-prone workflow to be able to rescale and normalize automatically and independently.

**Comment by abdksyed** [comment]
The means and standard deviation for my entire custom dataset is:
mean = [0.5058152675628662, 0.5046470165252686, 0.47556382417678833]
mean_255 =  [0.5058152675628662*255, 0.5046470165252686*255, 0.47556382417678833*255]
std = [0.19492648541927338, 0.19461843371391296, 0.21560588479042053]
std_255 = [0.19492648541927338*255, 0.19461843371391296*255, 0.21560588479042053*255]

So my normalize should be 1) ops.Normalize(device= "gpu", mean = mean, std = std) 
                                      or 2)  ops.Normalize(device= "gpu", mean = mean_255, std = std_255)

**Comment by JanuszL** [comment]
The [normalize operator](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/supported_ops.html#nvidia.dali.ops.Normalize) should address this feature request.

**Comment by mzient** [comment]
If you want to use operator `Normalize` with non-scalar mean/stddev, it has to be provided as a data node.
Specifying three values as a 1D array is not enough, since any axis can be subject to normalization.

The correct usage is:
```Python
mean = np.array([[[0.5058152675628662, 0.5046470165252686, 0.47556382417678833]]])
std = np.array([[[0.19492648541927338, 0.19461843371391296, 0.21560588479042053]]])

norm_op = ops.Normalize(device="gpu")

images = decoder_op(image_files)
normalized = norm_op(images, mean=nvidia.dali.types.Constant(mean * 255), std=nvidia.dali.types.Constant(std * 255))
```
For channel-first layout, you'd specify the mean/stddev as:
```Python
mean = np.array([[[0.5058152675628662]], [[0.5046470165252686]], [[0.47556382417678833]]])
std = np.array([[[0.19492648541927338]], [[0.19461843371391296]], [[0.21560588479042053]]])
```

**Comment by mzient** [comment]
...of course, if you want to normalize pixel values in your image (not using predefined values), youd simply use
```Python
ops.Normalize(device="gpu")  # normalize all channels together
```
Assuming you have a HWC or CHW layout and you want to normalize channels separately, you'd use
```Python
ops.Normalize(device="gpu", axis_names="HW")
```

You can also go halfway between per-image and pre-baked and normalize the values for the whole batch
```Python
ops.Normalize(device="gpu", axis_names="HW", batch=True)
```
This will calculate mean/std for each channel, but will use pixels from all images in the batch.

**Comment by abdksyed** [comment]
@mzient 
Thanks for your help. I understood that I have to give a 3d array and convert to DataNode type and not as list.
Also I was working around with CropMirrorNormalize, without giving any Crop and Mirror arguments, so that I can also convery the structure from NHWC to NCHW (N-Batch Size, H-Height, W-Width, C-Channels), so cal also provide the output type argument to normalize function?

Also I am little confused, the mean values and std value which I have provided as list is channel wise mean of the entire dataset, and would like to normalize the entire dataset with those values channel wise, like the first value of mean/std for first channel of image and 2nd value for 2nd channel and 3rd for 3rd.

So which of these method mentioned by you should be suitable for applying normalize channel wise
1) mean = np.array([[[0.5058152675628662, 0.5046470165252686, 0.47556382417678833]]])
2) mean = np.array([[[0.5058152675628662]], [[0.5046470165252686]], [[0.47556382417678833]]])

**Comment by mzient** [comment]
@abdksyed 
If you're limited to images and use pre-baked mean/stddev, then indeed you can use use CropMirrorNormalize. It has a dtype argument, but only supports float and float16 data types. If you also want to convert to CHW layout in one go, then I'd say it's the way to go.

`Normalize` is much more flexible when it comes to kinds of data it can process. It doesn't really have a notion of "channels" - think about it as the following numpy code:
```Python
normalized = (x - mean) / std
```
with the restriction that the size of mean and std must match.
So, the representation of mean and std will depend on the data layout. For HWC data it would be the first one (`mean = np.array([[[0.5058152675628662, 0.5046470165252686, 0.47556382417678833]]])`), for CHW the second (`mean = np.array([[[0.5058152675628662]], [[0.5046470165252686]], [[0.47556382417678833]]])`)
